##  Segment and Cluster Neighbourhoods in Toronto
### IBM/Coursera Applied Datascience Capstone

This Project Notebook assignment is Part 1 of 3 for Week 3: Applied Datascience Capstone Course assignment. 


Load url into data objects for analysis: 

Step 1: Build code to scrape following wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the Postal Code, Borough, and Neighborhood data.

In [1]:
# Load url into data objects for analysis: 
import pandas as pd #library for data analysis
data = pd.read_html("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")

Wrangle and clean data:

Step 2: Evaluate each of the data table objects in turn that we created during url loading.

Step 3: Designate primary dataframe and populate with required data objects loaded from url.

In [2]:
# Explore first few rows of our initial data[0] table
# Create df, primary dataframe, merged from data[0].
# Designate df as primary dataframe. 

df = data[0]
df.head(5)

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

####  Summary: 
- The first dataframe object data[0] created by html reader has the 3 column names we require, and is populated with 180 rows of data.
- data[1] and data[2], our second and third imported dataframe objects, were both created during scraping by read_html and evaluated to see if any data is required of them. Both of the data sets contain Canada province/state higher level information, not in scope for our Toronto City level model, therefore are not used.
- Steps 1 thru 3 are complete.

- Step 4: Ensure Borough (town) cells with an assigned borough are processed. Cells where borough = 'Not assigned' are ignored.

In [3]:
#Identify cell counts for each Borough. 
# We see 77 Borough cells where name is 'Not Assigned
print(df.Borough.value_counts()) 

Not assigned        77
North York          24
Downtown Toronto    19
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
East Toronto         5
York                 5
East York            5
Mississauga          1
Name: Borough, dtype: int64


In [4]:
# Remove 'Not assigned" count data on Borough
df = df[df['Borough'] != 'Not assigned']
print(df.Borough.value_counts())

North York          24
Downtown Toronto    19
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
East York            5
East Toronto         5
York                 5
Mississauga          1
Name: Borough, dtype: int64


#### Summary:
- 77 row counts found above identified where Borough column has 'Not Assigned' row value.
- After applying Step 4, 77 'Not assigned' rows are successfully removed for Borough column
- Step 4 is complete.

Step 5: When two or more Neighborhoods share the same postal code, combine each Neighbourhood to a single neighborhood line comma delimited.

In [5]:
# I first perform a count sum of any df column duplicates by creating a lambda function. 
dup_chk = df.apply(lambda col: col.duplicated()).sum()
print(dup_chk)
print(df.head(5))

Postal Code       0
Borough          93
Neighbourhood     4
dtype: int64
  Postal Code           Borough                                Neighbourhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government


#### Summary: 
1. The dup_chk lambda function above returned a 0 sum of Postal Code column duplicates, therefore no condition of two or more Neighborhoods sharing same Postal Code value exists within dataframe. 
2. Neighbourhood column above shows comma delimited Neighbourhoods already getting combined to a single line as desired, where multiple neighbourhoods share same Postal Code.
3. No problem occurences found, already working. Step 5 is complete.

Step 6: If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.

In [6]:
#List all rows where Neighbourhood = 'Not assigned'. 
Num_Neighbourhood_Not_Assigneds = df[df['Neighbourhood'] == 'Not assigned']
Num_Neighbourhood_Not_Assigneds.head(5)

Empty DataFrame
Columns: [Postal Code, Borough, Neighbourhood]
Index: []

#### Summary: 
1. No applicable 'Not assigned" Neighborhood cases found where Neighbourhood cells needed to be filled with Borough value.
3. Step 6 reviewed for eligible 'Not assigned' cases where none were found. No value assignment required. Complete.

Step 7: Use .shape method to print number of rows of dataframe

In [7]:
print(df.shape, 'rows x cols')

(103, 3) rows x cols


Save final dataframe to csv file to make available for next phase part 2 import.

In [8]:
#Create a Wk3:PostalDataPart1.csv file we will use for Wk3: Part 2.
df.to_csv('Wk3PostalDataPart1.csv', index=False)

#Test csv load
df2 = pd.read_csv('./Wk3PostalDataPart1.csv')
df2.head()

Postal Code           Borough                                Neighbourhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

#### Summary: 
- Net of 103 net rows of data, after 77 borough 'Not Assigned' rows were removed from initial 180 rows reported during initial step.
- Dataframe is ready for integration with next Part 2 assignment.

### This concludes Part 1 of 3 for Week 3: Applied Datascience Capstone Course assignment. 